Assets/Scripts/UIManager.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.UI;
+using UnityEngine.EventSystems;
+
+public class UIManager : MonoBehaviour
+{
+    public GameObject placement_object_prefab;
+
+    Ray mouse_ray;
+    RaycastHit hit;
+
+    GameObject tile;
+
+    GraphicRaycaster m_Raycaster;
+    PointerEventData m_PointerEventData;
+    EventSystem m_EventSystem;
+
+    public GameObject focused_placeholder;
+
+    private void Start()
+    {
+    }
+
+    private void Update()
+    {
+        ManageMouseHighlighter();
+        ClickManager();
+    }
+
+    void ManageMouseHighlighter()
+    {
+        mouse_ray = Camera.main.ScreenPointToRay(Input.mousePosition);
+
+        if (Physics.Raycast(mouse_ray, out hit))
+        {
+            if (hit.collider.gameObject.tag == "placeholder")
+            {
+                focused_placeholder = hit.collider.gameObject;
+            }
+        }
+        else
+        {
+            focused_placeholder = null;
+        }
+    }
+
+    void ClickManager()
+    {
+        if (Input.GetKeyDown(KeyCode.Mouse0))
+        {
+            if(focused_placeholder && placement_object_prefab)
+            {
+                // place the object prefab in the placeholder.
+                if(focused_placeholder.GetComponent<Placeholder>().type == placement_object_prefab.GetComponent<Placeable>().type)
```

Comment:
1. Use https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnMouseOver.html
2. Use https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnMouseDown.html
3. Let objects maintain their own behavior - if a placeholder is going to be changed, the component doing the changing should be on that object

Reply:
What component is doing the 'changing' here though? Isn't it the user click? So I decided a UI Manager object should handle user click interactions. How would you propose item #3 look?